harness/determined/core/_checkpoint.py

```diff
@@ -58,42 +159,132 @@ def __init__(
+        # far more complicated than having two codepaths.
+        if not shard:
+            if self._dist.rank != 0:
+                raise RuntimeError(
+                    f"cannot call .upload(shard=False) from non-chief worker "
+                    f"(rank={self._dist.rank})"
+                )
+            if ckpt_dir is None:
+                raise RuntimeError(
+                    "cannot call .upload(ckpt_dir=None, shard=False), which would result in doing "
+                    "nothing at all"
+                )
+            return self._upload_single(ckpt_dir, metadata)
+        else:
+            storage_id = None
+            if self._dist.rank == 0:
+                storage_id = str(uuid.uuid4())
+            storage_id = self._dist.broadcast(storage_id)
 
-        if self._dist.rank != 0:
-            raise RuntimeError(
-                "cannot call CheckpointContext.upload() from non-chief worker "
-                f"(rank={self._dist.rank})"
-            )
-
-        ckpt_dir = os.fspath(ckpt_dir)
+            assert storage_id
+            return self._upload_sharded(ckpt_dir, storage_id, metadata)
 
+    def _upload_single(self, ckpt_dir: str, metadata: Optional[Dict[str, Any]] = None) -> str:
         storage_id = str(uuid.uuid4())
         resources = self._storage_manager._list_directory(ckpt_dir)
 
-        # add metadata pre-upload but without counting it among resources
+        # Add metadata pre-upload but without counting it among resources.
         self._write_metadata_file(ckpt_dir, metadata or {})
 
         self._storage_manager.upload(src=ckpt_dir, dst=storage_id)
         self._report_checkpoint(storage_id, resources, metadata)
         return storage_id
 
+    def _upload_sharded(
+        self, ckpt_dir: Optional[str], storage_id: str, metadata: Optional[Dict[str, Any]] = None
+    ) -> str:
+        ckpt_dir_mask = self._dist.allgather(ckpt_dir is not None)
+        if not any(ckpt_dir_mask):
+            raise RuntimeError(
+                "cannot call .upload(ckpt_dir=None, shard=True), from all ranks; "
+                "at least one rank must have a valid ckpt_dir"
+            )
+
+        # Deconflict locally-shared directories; if every worker uploads /tmp/ckpt, then only
+        # the lowest rank on each node will actually upload this directory.
+        if ckpt_dir is None:
+            file_uid = None
+        else:
+            st = os.stat(ckpt_dir)
+            file_uid = (st.st_dev, st.st_ino)
+        all_file_uids = self._dist.allgather(file_uid)
+        # Decide if our rank is the lowest rank trying to upload this ckpt_dir.
+        want_upload = file_uid and all_file_uids.index(file_uid) == self._dist.rank
+
+        # Decide what we are going to upload.
+        if want_upload:
+            assert ckpt_dir
+            resources = self._storage_manager._list_directory(ckpt_dir)
+        else:
+            resources = {}
+
+        # Merge resources, detect conflicts
```

Comment:
I know that some of these are my errors, and not yours, but we generally try to keep comments in our code as starting with capital letters, and ending in periods.

Can you clean up the comments in this PR to match that?

If one or two slip through that's not the end of the world, but it's nice when the code mostly looks uniform.